presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt

```diff
@@ -1,61 +1,93 @@
+import com.exallium.mvvmcleanexample.presentation.utils.bindable
+import com.exallium.mvvmcleanexample.presentation.utils.delegates.EcoDisposableProperty
+import com.exallium.mvvmcleanexample.presentation.utils.distinctObservable
+import com.exallium.mvvmcleanexample.presentation.utils.ecoDisposable
+import com.exallium.mvvmcleanexample.presentation.utils.notifiable.BaseNotifiableObservable
+import com.exallium.mvvmcleanexample.presentation.utils.notifiable.NotifiableObservable
 import com.jakewharton.rxrelay2.BehaviorRelay
 import com.jakewharton.rxrelay2.PublishRelay
-import io.reactivex.Observable
-import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.disposables.Disposable
 
-class UserEditViewModel(userEditModel: UserEditModel,
-                        private val router: UiRouter) {
+class UserEditViewModel(
+        private val userEditModel: UserEditModel,
+        private val router: UiRouter,
+        private val firstNameRelay: BehaviorRelay<String> = BehaviorRelay.create(),
+        private val lastNameRelay: BehaviorRelay<String> = BehaviorRelay.create(),
+        private val savedUserRelay: PublishRelay<User> = PublishRelay.create(),
+        ecoDisposableProperty: EcoDisposableProperty = ecoDisposable(null),
+        notifiableObservable: NotifiableObservable = BaseNotifiableObservable())
+    : NotifiableObservable by notifiableObservable {
 
-    val firstNameError = ObservableField<String>("")
-    val lastNameError = ObservableField<String>("")
+    init {
+        setDelegator(this)
+    }
 
-    private val compositeDisposable = CompositeDisposable()
+    /* symbol meaning:
+    >> : bind direction left to right, future changes are propagated
+    << : bind direction right to left, future changes are propagated
+    >| : bind direction left to right, future changes are not propagated
+    |< : bind direction right to left, future changes are not propagated
+     */
 
-    private val clickRelay = PublishRelay.create<User>()
-    private val firstNameRelay = BehaviorRelay.create<String>()
-    private val lastNameRelay = BehaviorRelay.create<String>()
+    // view |< >> view model
+    var firstName: String by distinctObservable(
+            initialValue = "",
+            onChange = firstNameRelay::accept)
 
-    init {
-        compositeDisposable.add(userEditModel.results(clickRelay, firstNameRelay, lastNameRelay)
-                .subscribe(this::handleResult, this::handleError))
-    }
+    // view |< >> view model
+    var lastName: String by distinctObservable(
+            initialValue = "",
+            onChange = lastNameRelay::accept)
 
-    fun setSaveClicks(obs: Observable<Unit>) {
-        compositeDisposable.add(obs
-                .map { User(firstNameRelay.value, lastNameRelay.value) }
-                .subscribe(clickRelay))
-    }
+    // view << view model
+    @get:Bindable
+    var firstNameError: String by bindable("")
+
```

Comment:
```
What I don't like about ObservableField and the primitive counterparts is that they expose public setters on the properties of the view model that are only intended to be observed by the view, where the state is managed internally by the view model.
```

If the View (and I guess the activity) are the only things getting access to the ViewModel, I'm not sure how relevant it is that they're public.